### ISSUE TITLE
-release in Scala doesn't expose (some) internal Java classes

### ISSUE THREAD

**Issue by ijuma** [scala/bug#10817]
Some may consider this a feature, but it differs from Java's behaviour. In Kafka, we have been using Java's `--release 7` for a while with no issues. I tried to change the build to do the same for Scala (with 2.12.5) and I get compilation errors like:

> /Users/ijuma/src/kafka/core/src/main/scala/kafka/Kafka.scala:23: not found: object sun
> import sun.misc.{Signal, SignalHandler}
>        ^
> /Users/ijuma/src/kafka/core/src/main/scala/kafka/Kafka.scala:59: not found: type SignalHandler
>     val jvmSignalHandlers = new ConcurrentHashMap[String, SignalHandler]().asScala
>                                                           ^
> /Users/ijuma/src/kafka/core/src/main/scala/kafka/Kafka.scala:60: not found: type SignalHandler
>     val handler = new SignalHandler() {
>                       ^
> /Users/ijuma/src/kafka/core/src/main/scala/kafka/Kafka.scala:61: not found: type Signal
>       override def handle(signal: Signal) {
>                                   ^
> /Users/ijuma/src/kafka/core/src/main/scala/kafka/Kafka.scala:67: not found: value Signal
>       val oldHandler = Signal.handle(new Signal(signalName), handler)
>                        ^
> /Users/ijuma/src/kafka/core/src/main/scala/kafka/Kafka.scala:67: not found: type Signal
>       val oldHandler = Signal.handle(new Signal(signalName), handler)

The Kafka PR: https://github.com/apache/kafka/pull/4837/files

**Comment by ijuma** [comment]
This is actually not true, sorry for the noise. When I moved the code to Java, it also failed to compile even though these classes were excluded from http://openjdk.java.net/jeps/260.

**Comment by ijuma** [comment]
Unfortunately, I found another issue when using `-release 8` and Java 10, but not yet sure how to minimise it:

```text
/Users/ijuma/src/kafka/core/src/main/scala/kafka/Kafka.scala:72: Class sun.reflect.CallerSensitive not found - continuing with a stub.
      Runtime.getRuntime().addShutdownHook(new Thread("kafka-shutdown-hook") {
```
But `Kafka` never uses that annotation:

https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/Kafka.scala#L88 (the line number is a bit different because I have some local changes).

After I commented out that line, I got a similar error somewhere else:

```text
/Users/ijuma/src/kafka/core/src/main/scala/kafka/admin/AdminClient.scala:58: Class sun.reflect.CallerSensitive not found - continuing with a stub.
  val networkThread = new KafkaThread("admin-client-network-thread", new Runnable {
```

https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/admin/AdminClient.scala#L58

This could be related to the fact that `CallerSensitive` was moved from `sun.reflect` in Java 8 to `jdk.internal.reflect` in Java 9/10.

Any ideas with regards to minimisation are welcome. The obvious case didn't work.
